statsmodels/genmod/families/family.py

```diff
@@ -1432,3 +1432,276 @@ def resid_anscombe(self, endog, mu):
+    `GLM` class.
+    """
+    links = [L.log, L.Power]
+    # TODO: add the ability to use the power links with an if test
+    # similar to below
+    variance = V.Power
+    safe_links = [L.log, L.Power]
+
+    def __init__(self, link=None, var_power=1., link_power=0):
+        self.var_power = var_power
+        self.link_power = link_power
+        self.variance = V.Power(power=var_power * 1.)
+        if link_power != 0 and not ((link is L.Power) or (link is None)):
+            msg = 'link_power of {} not supported specified link'
+            msg = msg.format(link_power)
+            raise ValueError(msg)
+        if (link_power == 0) and ((link is None) or (link is L.Log)):
+            self.link = L.log()
+        elif link_power != 0:
+            self.link = L.Power(power=link_power * 1.)
+        else:
+            self.link = link()
+
+    def _clean(self, x):
+        """
+        Helper function to trim the data so that is in (0,inf)
+
+        Notes
+        -----
+        The need for this function was discovered through usage and its
+        possible that other families might need a check for validity of the
+        domain.
+        """
+        return np.clip(x, 0, np.inf)
+
+    def deviance(self, endog, mu, freq_weights=1., scale=1.):
+        r"""
+        Returns the value of the deviance function.
+
+        Parameters
+        -----------
+        endog : array-like
+            Endogenous response variable
+        mu : array-like
+            Fitted mean response variable
+        freq_weights : array-like
+            1d array of frequency weights. The default is 1.
+        scale : float, optional
+            An optional scale argument. The default is 1.
+
+        Returns
+        -------
+        deviance : float
+            Deviance function as defined below
+
+        Notes
+        -----
+        When :math:`p = 1`,
+
+        .. math::
+
+            resid\_dev_i = \mu
+
+        when :math:`endog = 0` and
+
+        .. math::
+
+            resid\_dev_i = \mu - endog * (1 + \log(\mu / endog))
+
+        otherwise.
+
+        When :math:`p = 2`,
+
+        .. math::
+
+            resid\_dev_i = -1 + (endog / \mu) + \log(\mu / endog)
+
+        For all other p,
+
+        .. math::
+
+            resid\_dev_i = endog ^{2 - p} / ((1 - p) * (2 - p)) -
+                           endog * \mu ^{1 - p} / (1 - p) + \mu ^{2 - p} /
+                           (2 - p)
+
+        Once :math:`resid\_dev_i` is calculated, then calculate deviance as
+
+        .. math::
+
+            D = \sum{2 * freq\_weights * resid\_dev_i}
+        """
+        p = self.var_power
+        if p == 1:
+            dev = np.where(endog == 0,
+                           2. * mu,
+                           mu - endog * (1 + np.log(mu / endog)))
+        elif p == 2:
+            endog1 = np.clip(endog, FLOAT_EPS, np.inf)
+            dev = -1 + (endog / mu) + np.log(mu / endog1)
```

Comment:
Another possible change to reduce code duplication.
deviance and deviance residuals are almost the same except for the sum and the sign.
One possibility is to add an internal method that calculates the core computation for each observation, and then let deviance just calculate the weighted sum, and resid_dev calculates the signed sqrt.